<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [2]:
import piplite
await piplite.install(['folium'])
await piplite.install(['pandas'])

In [3]:
import folium
import pandas as pd

In [4]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [5]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

In [6]:
spacex_df.head(5)

Flight Number        Date Time (UTC) Booster Version  Launch Site  \
0              1  2010-06-04   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1              2  2010-12-08   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2              3  2012-05-22    7:44:00  F9 v1.0  B0005  CCAFS LC-40   
3              4  2012-10-08    0:35:00  F9 v1.0  B0006  CCAFS LC-40   
4              5  2013-03-01   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  Payload Mass (kg)  \
0               Dragon Spacecraft Qualification Unit                0.0   
1  Dragon demo flight C1, two CubeSats,  barrel o...                0.0   
2                             Dragon demo flight C2+              525.0   
3                                       SpaceX CRS-1              500.0   
4                                       SpaceX CRS-2              677.0   

       Orbit         Customer        Landing Outcome  class        Lat  \
0        LEO           SpaceX  Failure   (parachute)      0  28.562302   
1  LEO (ISS)  NASA (COTS) NRO  Failure   (parachute)      0  28.562302   
2  LEO (ISS)      NASA (COTS)             No attempt      0  28.562302   
3  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   
4  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   

        Long  
0 -80.577356  
1 -80.577356  
2 -80.577356  
3 -80.577356  
4 -80.577356

In [7]:
data_test = spacex_df.groupby('Launch Site')['class'].sum().reset_index()
data_test.head(5)

Launch Site  class
0   CCAFS LC-40      7
1  CCAFS SLC-40      3
2    KSC LC-39A     10
3   VAFB SLC-4E      4

In [8]:
#entered_site = 'CCAFS LC-40'
entered_site = 'CCAFS SLC-40'
#entered_site = 'KSC LC-39A'
#entered_site = 'VAFB SLC-4E'
data_test_sel = spacex_df[spacex_df['Launch Site']==entered_site]
data_test_sel.head(20)

Flight Number        Date Time (UTC) Booster Version   Launch Site  \
49             45  2017-12-15   15:36:00  F9 FT  B1035.2  CCAFS SLC-40   
50             47  2018-01-08    1:00:00   F9 B4 B1043.1  CCAFS SLC-40   
51             48  2018-01-31   21:25:00  F9 FT  B1032.2  CCAFS SLC-40   
52             50  2018-03-06    5:33:00     F9 B4 B1044  CCAFS SLC-40   
53             52  2018-04-02   20:30:00  F9 B4  B1039.2  CCAFS SLC-40   
54             53  2018-04-18   22:51:00   F9 B4 B1045.1  CCAFS SLC-40   
55             56  2018-06-04    4:45:00  F9 B4  B1040.2  CCAFS SLC-40   

                                         Payload  Payload Mass (kg)  \
49                                 SpaceX CRS-13            2205.00   
50                                          Zuma            3696.65   
51                             GovSat-1 / SES-16            4230.00   
52                        Hispasat 30W-6  PODSat            6092.00   
53                                 SpaceX CRS-14            2647.00   
54  Transiting Exoplanet Survey Satellite (TESS)             362.00   
55                                        SES-12            5384.00   

        Orbit             Customer       Landing Outcome  class        Lat  \
49  LEO (ISS)           NASA (CRS)  Success (ground pad)      1  28.563197   
50        LEO     Northrop Grumman  Success (ground pad)      1  28.563197   
51        GTO                  SES    Controlled (ocean)      0  28.563197   
52        GTO  Hispasat  NovaWurks            No attempt      0  28.563197   
53  LEO (ISS)           NASA (CRS)            No attempt      0  28.563197   
54        HEO           NASA (LSP)  Success (drone ship)      1  28.563197   
55        GTO                  SES            No attempt      0  28.563197   

        Long  
49 -80.57682  
50 -80.57682  
51 -80.57682  
52 -80.57682  
53 -80.57682  
54 -80.57682  
55 -80.57682

In [9]:
data_test_sel['class'].astype('float64')
data_test_sel.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7 entries, 49 to 55
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Flight Number      7 non-null      int64  
 1   Date               7 non-null      object 
 2   Time (UTC)         7 non-null      object 
 3   Booster Version    7 non-null      object 
 4   Launch Site        7 non-null      object 
 5   Payload            7 non-null      object 
 6   Payload Mass (kg)  7 non-null      float64
 7   Orbit              7 non-null      object 
 8   Customer           7 non-null      object 
 9   Landing Outcome    7 non-null      object 
 10  class              7 non-null      int64  
 11  Lat                7 non-null      float64
 12  Long               7 non-null      float64
dtypes: float64(3), int64(2), object(8)
memory usage: 560.0+ bytes


In [10]:
data_test_spacex_df = spacex_df.groupby(['Launch Site', 'class']).agg(
    sum_class=('class', 'sum'),
    avg_class=('class', 'mean'),
    count_class=('class', 'count')
)
#data
data_test_spacex_df

sum_class  avg_class  count_class
Launch Site  class                                   
CCAFS LC-40  0              0        0.0           19
             1              7        1.0            7
CCAFS SLC-40 0              0        0.0            4
             1              3        1.0            3
KSC LC-39A   0              0        0.0            3
             1             10        1.0           10
VAFB SLC-4E  0              0        0.0            6
             1              4        1.0            4

In [11]:
data_test_sel2 = data_test_sel.groupby('class').agg(
    sum_class=('class', 'sum'),
    avg_class=('class', 'mean'),
    count_class=('class', 'count')
)
#data_test_sel2 = spacex_df[['Launch Site', 'class']].groupby(['class'])['Launch Site'].mean().reset_index(name='avg sucess')
data_test_sel2

sum_class  avg_class  count_class
class                                   
0              0        0.0            4
1              3        1.0            3

In [12]:
#data_test_sel2 = data_test_sel.groupby('class')['Launch Site'].mean().reset_index()
#data_test_sel2 = data_test_sel.groupby(['Launch Site', 'class']).sum().reset_index()
data_test_sel2 = data_test_sel.groupby(['Launch Site', 'class']).agg(
    sum_class=('class', 'sum'),
    avg_class=('class', 'mean'),
    count_class=('class', 'count')
)
#data_test_sel2 = spacex_df[['Launch Site', 'class']].groupby(['class'])['Launch Site'].mean().reset_index(name='avg sucess')
data_test_sel2

sum_class  avg_class  count_class
Launch Site  class                                   
CCAFS SLC-40 0              0        0.0            4
             1              3        1.0            3

Now, you can take a look at what are the coordinates for each site.


In [13]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [14]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [15]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [16]:
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

In [17]:
launch_sites_df['Launch Site'].unique()

array(['CCAFS LC-40', 'CCAFS SLC-40', 'KSC LC-39A', 'VAFB SLC-4E'],
      dtype=object)

In [18]:
for i in range(0, launch_sites_df.shape[0]):
    print(f"{launch_sites_df['Launch Site'].loc[i]} {launch_sites_df['Lat'].loc[i]} {launch_sites_df['Long'].loc[i]}")

CCAFS LC-40 28.56230197 -80.57735648
CCAFS SLC-40 28.56319718 -80.57682003
KSC LC-39A 28.57325457 -80.64689529
VAFB SLC-4E 34.63283416 -120.6107455


In [19]:
import numpy as np

In [20]:
for i, column in enumerate(launch_sites_df):
    print(i, np.asarray(launch_sites_df[column]))

0 ['CCAFS LC-40' 'CCAFS SLC-40' 'KSC LC-39A' 'VAFB SLC-4E']
1 [28.56230197 28.56319718 28.57325457 34.63283416]
2 [ -80.57735648  -80.57682003  -80.64689529 -120.6107455 ]


In [21]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
for i in range(0, launch_sites_df.shape[0]):
    coordinate = [launch_sites_df['Lat'].loc[i], launch_sites_df['Long'].loc[i]]
    popup = launch_sites_df['Launch Site'].loc[i]
    label = launch_sites_df['Launch Site'].loc[i]
    circle = folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(popup))
    marker = folium.map.Marker(
    coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % label,
        )
    )
#    print(f"{launch_sites_df['Launch Site'].loc[i]} {launch_sites_df['Lat'].loc[i]} {launch_sites_df['Long'].loc[i]}")
    print(f"{label} {coordinate}")
    site_map.add_child(circle)
    site_map.add_child(marker)
site_map

CCAFS LC-40 [28.56230197, -80.57735648]
CCAFS SLC-40 [28.56319718, -80.57682003]
KSC LC-39A [28.57325457, -80.64689529]
VAFB SLC-4E [34.63283416, -120.6107455]


The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [77]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [78]:
marker_cluster = MarkerCluster()


*TODO:* Create a new column in `spacex_df` dataframe called `marker_color` to store the marker colors based on the `class` value


In [79]:
# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
spacex_df['marker_color'] = spacex_df['class'].apply(
    lambda x: 'green' if x == 1 else 'red'
    )
spacex_df.tail(10)

Launch Site        Lat       Long  class marker_color
46    KSC LC-39A  28.573255 -80.646895      1        green
47    KSC LC-39A  28.573255 -80.646895      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
49  CCAFS SLC-40  28.563197 -80.576820      1        green
50  CCAFS SLC-40  28.563197 -80.576820      1        green
51  CCAFS SLC-40  28.563197 -80.576820      0          red
52  CCAFS SLC-40  28.563197 -80.576820      0          red
53  CCAFS SLC-40  28.563197 -80.576820      0          red
54  CCAFS SLC-40  28.563197 -80.576820      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [80]:
for index, record in spacex_df.iterrows():
    if index < 2:
        print(f"index: {index}\nrecord: {record}\n")

index: 0
record: Launch Site     CCAFS LC-40
Lat               28.562302
Long             -80.577356
class                     0
marker_color            red
Name: 0, dtype: object

index: 1
record: Launch Site     CCAFS LC-40
Lat               28.562302
Long             -80.577356
class                     0
marker_color            red
Name: 1, dtype: object



In [81]:
for index, record in spacex_df.iterrows():
    if index < 2:
        print(f"index: {index}\nrecord: {record.iloc[0:3]}\n")

index: 0
record: Launch Site    CCAFS LC-40
Lat              28.562302
Long            -80.577356
Name: 0, dtype: object

index: 1
record: Launch Site    CCAFS LC-40
Lat              28.562302
Long            -80.577356
Name: 1, dtype: object



In [82]:
for index, record in spacex_df.iterrows():
    if index < 2:
        print(f"Index: {index}\n\
              Launch Site: {record['Launch Site']}, Lat: {record['Lat']}, Long: {record['Long']}\n")

Index: 0
              Launch Site: CCAFS LC-40, Lat: 28.56230197, Long: -80.57735648

Index: 1
              Launch Site: CCAFS LC-40, Lat: 28.56230197, Long: -80.57735648



In [83]:
count = 0
for index, record in spacex_df.iterrows():
    if record['Launch Site'] == 'CCAFS SLC-40':
        count += 1
        print(f"count: {count: 02d}: {record['Launch Site']}, {record['marker_color']}")

count:  1: CCAFS SLC-40, green
count:  2: CCAFS SLC-40, green
count:  3: CCAFS SLC-40, red
count:  4: CCAFS SLC-40, red
count:  5: CCAFS SLC-40, red
count:  6: CCAFS SLC-40, green
count:  7: CCAFS SLC-40, red


In [84]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    # TODO: Create and add a Marker cluster to the site map
    coordinate = [record['Lat'], record['Long']]
    label = record['Launch Site']
    marker = folium.map.Marker(
    coordinate,
    # Create an icon as a text label
        icon=folium.Icon(color='white', icon_color=record['marker_color'],
        #icon=DivIcon(
        #icon_size=(20,20),
        #icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % label,
        )
    )    
    marker_cluster.add_child(marker)

site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [85]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [86]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


CCAFS LC-40 [28.56230197, -80.57735648]
CCAFS SLC-40 [28.56319718, -80.57682003]
KSC LC-39A [28.57325457, -80.64689529]
VAFB SLC-4E [34.63283416, -120.6107455]

In [87]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
# Launch Site: KSC LC-39A
launch_site_lat = 28.57325457
launch_site_lon = -80.64689529
# Launch Site: CCAFS SLC-40 
launch_site_lat = 28.56319718
launch_site_lon = -80.57682003
coordinate_launch_site = [launch_site_lat, launch_site_lon]
##coastline_lat = 28.56367 #TO DO
##coastline_lon = -80.57163 #TO DO
#coastline_lat = 28.5628
#coastline_lon = -80.5679
coastline_lat = 28.56327
coastline_lon = -80.56799
coordinate_coastline = [coastline_lat, coastline_lon]
distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

In [88]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
distance_marker = folium.Marker(
    coordinate_coastline,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline),
        )
    )
marker_cluster.add_child(distance_marker)

site_map

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [89]:
coordinates = [coordinate_coastline, coordinate_launch_site]

In [90]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
lines=folium.PolyLine(locations=coordinates, weight=1)
site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [1]:
launch_site_lat = 28.56319718
launch_site_lon = -80.57682003

In [92]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
coordinate_city = [28.38793,-80.60544]
coordinates_to_city = [coordinate_launch_site, coordinate_city]
lines_city=folium.PolyLine(locations=coordinates_to_city, weight=1)
site_map.add_child(lines_city)

In [ ]:
# find coordinate of the closet city
#coordinate_launch_site = [launch_site_lat, launch_site_lon]
#coastline_lat = 28.56327
#coastline_lon = -80.56799
#coordinate_city = [28.38793,-80.60544]
city_lat = coordinate_city[0]
city_lon = coordinate_city[1]
distance_city = calculate_distance(launch_site_lat, launch_site_lon, city_lat, city_lon)

In [ ]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
distance_marker_city = folium.Marker(
    coordinate_city,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_city),
        )
    )
marker_cluster.add_child(distance_marker_city)
#site_map

In [93]:
coordinate_railway = [launch_site_lat,launch_site_lon]
coordinates_to_railway = [coordinate_launch_site, coordinate_railway]
lines_railway=folium.PolyLine(locations=coordinates_to_railway, weight=1)
site_map.add_child(lines_railway)

In [ ]:
# find coordinate of the closet highway
#coordinate_launch_site = [launch_site_lat, launch_site_lon]
#coastline_lat = 28.56327
#coastline_lon = -80.56799
#coordinate_highway = [28.56326,-80.57077]
railway_lat = coordinate_railway [0]
railway_lon = coordinate_railway [1]
distance_railway = calculate_distance(launch_site_lat, launch_site_lon, railway_lat, railway_lon)

In [ ]:
# Create and add a folium.Marker on your selected closest highway point on the map
distance_marker_railway = folium.Marker(
    coordinate_railway,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_railway),
        )
    )
marker_cluster.add_child(distance_marker_railway)

site_map

In [94]:
coordinate_highway = [28.56339,-80.5708]
coordinate_highway = [28.56326,-80.57077]
coordinates_to_highway = [coordinate_launch_site, coordinate_highway]
lines_highway=folium.PolyLine(locations=coordinates_to_highway, weight=1)
site_map.add_child(lines_highway)

In [ ]:
# find coordinate of the closet highway
#coordinate_launch_site = [launch_site_lat, launch_site_lon]
#coastline_lat = 28.56327
#coastline_lon = -80.56799
#coordinate_highway = [28.56326,-80.57077]
highway_lat = coordinate_highway[0]
highway_lon = coordinate_highway[1]
distance_highway = calculate_distance(launch_site_lat, launch_site_lon, highway_lat, highway_lon)

In [ ]:
# Create and add a folium.Marker on your selected closest highway point on the map
distance_marker_highway = folium.Marker(
    coordinate_highway,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_highway),
        )
    )
marker_cluster.add_child(distance_marker_highway)

site_map

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)


<!--## Change Log--!>


<!--| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|--!>


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
